### PR TITLE
FIX KYLIN-4810

### PR DESCRIPTION
**Pull request by zhengshengjun**
## Proposed changes

Fix the problem of dictionary build error when encountering large values,  related issue link: https://issues.apache.org/jira/browse/KYLIN-4810

## Types of changes

What types of changes does your code introduce to Kylin?
_Put an `x` in the boxes that apply_

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update (if none of the other choices apply)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [x] I have create an issue on [Kylin's jira](https://issues.apache.org/jira/browse/KYLIN), and have described the bug/feature there in detail
- [ ] Commit messages in my PR start with the related jira ID, like "KYLIN-0000 Make Kylin project open-source"
- [x] Compiling and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If this change need a document change, I will prepare another pr against the `document` branch
- [ ] Any dependent changes have been merged

## Further comments

If this is a relatively large or complex change, kick off the discussion at user@kylin or dev@kylin by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by hit-lacus** [comment]
+1

**Comment by hit-lacus** [comment]
@bigxiaochu Could you please help to review this patch ?

File: core-dictionary/src/main/java/org/apache/kylin/dict/TrieDictionaryBuilder.java [hunk 1]
```diff
@@ -108,11 +112,13 @@ private void addValueR(Node node, byte[] value, int start) {
...
                 // if equals to current node, just mark end of value
-                node.isEndOfValue = true;
+                if (!isSplitValue) {
+                    node.isEndOfValue = true;
+                }
             } else {
                 // otherwise, split the current node into two
                 Node c = new Node(BytesUtil.subarray(node.part, i, n), node.isEndOfValue, node.children);
-                node.reset(BytesUtil.subarray(node.part, 0, i), true);
+                node.reset(BytesUtil.subarray(node.part, 0, i), isSplitValue? false : true);
```
Comment (guangxuCheng) [inline core-dictionary/src/main/java/org/apache/kylin/dict/TrieDictionaryBuilder.java:121]: `isSplitValue? false : true` can be simplified to `!isSplitValue`

Comment (guangxuCheng) [inline core-dictionary/src/main/java/org/apache/kylin/dict/TrieDictionaryBuilder.java:121]: `isSplitValue? false : true` can be simplified to `!isSplitValue`

File: core-dictionary/src/main/java/org/apache/kylin/dict/TrieDictionaryBuilder.java [hunk 2]
```diff
@@ -121,7 +127,7 @@ private void addValueR(Node node, byte[] value, int start) {
         // if partially matched the current, split the current node, add the new value, make a 3-way
         if (i < n) {
             Node c1 = new Node(BytesUtil.subarray(node.part, i, n), node.isEndOfValue, node.children);
-            Node c2 = new Node(BytesUtil.subarray(value, j, nn), true);
+            Node c2 = new Node(BytesUtil.subarray(value, j, nn), isSplitValue? false : true);
```
Comment (guangxuCheng) [inline core-dictionary/src/main/java/org/apache/kylin/dict/TrieDictionaryBuilder.java:130]: `isSplitValue? false : true` can be simplified to `!isSplitValue`

File: core-dictionary/src/main/java/org/apache/kylin/dict/TrieDictionaryBuilder.java [hunk 3]
```diff
@@ -152,10 +158,10 @@ else if (comp > 0)
         }
         if (found) {
             // found a child node matching the first byte, continue in that child
-            addValueR(node.children.get(mid), value, j);
+            addValueR(node.children.get(mid), value, j, isSplitValue);
         } else {
             // otherwise, make the value a new child
-            Node c = new Node(BytesUtil.subarray(value, j, nn), true);
+            Node c = new Node(BytesUtil.subarray(value, j, nn), isSplitValue ? false : true);
```
Comment (guangxuCheng) [inline core-dictionary/src/main/java/org/apache/kylin/dict/TrieDictionaryBuilder.java:164]: `isSplitValue? false : true` can be simplified to `!isSplitValue`

**Review by guangxuCheng — COMMENTED** [review]
Leave some tips and It's best to add some unit tests

File: core-dictionary/src/main/java/org/apache/kylin/dict/TrieDictionaryBuilder.java (hunk 3 above)
Comment (zhengshengjun) [reply core-dictionary/src/main/java/org/apache/kylin/dict/TrieDictionaryBuilder.java]: thanks for your reviewing and advice

**Comment by zhengshengjun** [comment]
i have added a test case and simplified the code, could please help to review this patch again ? @guangxuCheng

**Review by guangxuCheng — APPROVED** [review]
LGTM +1

**Review by hit-lacus — APPROVED** [review]
LGTM

**Comment by bigxiaochu** [comment]
+1， good job.

**Comment by bigxiaochu** [comment]
+1， good job.

**Comment by hit-lacus** [comment]
Merged in https://github.com/apache/kylin/pull/1486 , thanks @zhengshengjun .

**Pull request closed without merging**